packages/host/app/services/matrix-service.ts

```diff
@@ -228,13 +229,96 @@ export default class MatrixService extends Service {
       try {
         await this._client.startClient();
         await this.initializeRooms();
+
+        // TODO this is a temporary measure to prove that we can obtain a realm session.
+        // ultimately we need to figure out a better approach in terms of when/where we
+        // obtain sessions for realms that we care about.
+        await this.getRealmToken(new URL(ownRealmURL));
       } catch (e) {
         console.log('Error starting Matrix client', e);
         await this.logout();
       }
     }
   }
 
+  async getRealmToken(
+    realmURL: URL,
+  ): Promise<TokenClaims & { iat: number; exp: number }> {
+    let tokenRefreshPeriod = 5 * 60; // 5 minutes
+    let tokens = JSON.parse(localStorage.getItem('boxel-session') ?? '{}') as {
+      [realm: string]: TokenClaims & { exp: number; iat: number };
+    };
+    let token = tokens[realmURL.href];
+    if (token) {
+      let expiration = token.exp;
+      if (expiration - tokenRefreshPeriod > Date.now() / 1000) {
+        return token;
+      }
+    }
+    await this.createRealmSession(realmURL);
+    return await this.getRealmToken(realmURL);
+  }
+
+  private async createRealmSession(realmURL: URL) {
+    await this.ready;
+    if (!this.isLoggedIn) {
+      throw new Error(
+        `must be logged in to matrix before a realm session can be created`,
+      );
+    }
+
+    let initialResponse = await fetch(`${realmURL.href}_session`, {
+      method: 'POST',
+      headers: {
+        Accept: 'application/json',
+      },
+      body: JSON.stringify({
+        user: this.userId,
+      }),
+    });
+    let initialJSON = (await initialResponse.json()) as {
+      room: string;
+      challenge: string;
+    };
+    if (initialResponse.status !== 401) {
+      throw new Error(
+        `unexpected response from POST ${realmURL.href}_session: ${
+          initialResponse.status
+        } - ${JSON.stringify(initialJSON)}`,
+      );
+    }
+    let { room, challenge } = initialJSON;
+    if (!this.rooms.has(room)) {
+      await this.client.joinRoom(room);
+    }
+    await this.sendMessage(room, `auth-response: ${challenge}`);
+    let challengeResponse = await fetch(`${realmURL.href}_session`, {
+      method: 'POST',
+      headers: {
+        Accept: 'application/json',
+      },
+      body: JSON.stringify({
+        user: this.userId,
+        challenge,
+      }),
+    });
+    let jwtJSON = (await challengeResponse.json()) as { jwt: string };
+    if (!challengeResponse.ok) {
+      throw new Error(
+        `Could not authenticate with realm ${realmURL.href} - ${
+          challengeResponse.status
+        }: ${JSON.stringify(jwtJSON)}`,
+      );
+    }
+    let { jwt } = jwtJSON;
+    let sessionStr = localStorage.getItem('boxel-session') ?? '{}';
+    let session = JSON.parse(sessionStr);
+    session[realmURL.href] = jwt;
+    localStorage.setItem('boxel-session', JSON.stringify(session));
```

Comment:
@backspace this is what i'm thinking in terms of storing our JWT

Reply:
maybe this could delegate to a `SessionsService` that manages access to `localStorage` so we can preserve reactivity? Something like `persistSession(realmURL, session)`

Reply:
I think this is good for this PR though, I can include a change like that in #955

Reply:
sounds good